[WebGPU] Creating a ComputePipeline with a vertexShader module or similar should be an error

mwyrzykowski opened the pull request:
#### ae54cd2196fe0a8ccbe5796cc913a67b8ade8e93
<pre>
[WebGPU] Creating a ComputePipeline with a vertexShader module or similar should be an error
<a href="https://bugs.webkit.org/show_bug.cgi?id=265928">https://bugs.webkit.org/show_bug.cgi?id=265928</a>
&lt;radar://119234036&gt;

Reviewed by Tadeu Zagallo.

Creating a pipeline now allows the entryPoint to be omitted
and the only suitable default will be chosen.

Also, if we attempt to create a compute pipeline with a vertex
shader or similar incorrect combinations, that should be an error.

* Source/WebCore/Modules/WebGPU/GPUProgrammableStage.idl:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::Device::createShaderModule):
(WebGPU::ShaderModule::ShaderModule):
(WebGPU::ShaderModule::convertPipelineLayout):
(WebGPU::ShaderModule::defaultVertexEntryPoint const):
(WebGPU::ShaderModule::defaultFragmentEntryPoint const):
(WebGPU::ShaderModule::defaultComputeEntryPoint const):

Canonical link: <a href="https://commits.webkit.org/271652@main">https://commits.webkit.org/271652@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/825268ab278690ef4a342918642756adf71092fa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31683 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26475 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29675 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5057 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26487 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29351 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24934 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5554 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5681 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25962 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33021 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26560 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26399 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31927 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5661 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3834 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29723 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7322 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6951 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6162 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6182 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -152,6 +154,27 @@
+        for (auto& function : check.ast->functions()) {
+            if (!function.stage())
+                continue;
+            switch (*function.stage()) {
```
do we need to check that there is only one candidate?

mwyrzykowski replied:
oh that is a good point, yes I should leave the default empty if there is more than one candidate

mwyrzykowski committed:
[WebGPU] Creating a ComputePipeline with a vertexShader module or similar should be an error
https://bugs.webkit.org/show_bug.cgi?id=265928
<radar://119234036>

Reviewed by Tadeu Zagallo.

Creating a pipeline now allows the entryPoint to be omitted
and the only suitable default will be chosen.

Also, if we attempt to create a compute pipeline with a vertex
shader or similar incorrect combinations, that should be an error.

* Source/WebCore/Modules/WebGPU/GPUProgrammableStage.idl:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::Device::createComputePipeline):
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
* Source/WebGPU/WebGPU/ShaderModule.h:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::Device::createShaderModule):
(WebGPU::ShaderModule::ShaderModule):
(WebGPU::ShaderModule::convertPipelineLayout):
(WebGPU::ShaderModule::defaultVertexEntryPoint const):
(WebGPU::ShaderModule::defaultFragmentEntryPoint const):
(WebGPU::ShaderModule::defaultComputeEntryPoint const):

Canonical link: https://commits.webkit.org/271652@main
